### PR TITLE
Add matching allowed-alt-sxg for preloads.

### PR DESCRIPTION
**Pull request by twifkak**
For each same-origin `Link: rel=preload directive`, fetch the subresource, sign
it, compute its header-integrity, and construct a corresponding `Link:
rel=allowed-alt-sxg`. This allows subresource substitution [1], as required by
the Google SXG Cache for any preloads [2].

Header-integrity values are cached, since fetching and signing each subresource
is expensive. If there is an error in fetching or signing, then the error
status is also cached; this reduces the traffic to the origin caused by
preloading a resource that can't be signed.

To facilitate integration with the environment:
- Added a subresource_fetcher parameter which is an implementation of Fetcher,
  but may be different from the OCSP fetcher if desired.
- Created an HttpCache trait, and provided an implementation which wraps the
  service workers Cache API [3], as provided by Cloudflare Workers [4].
- Added a HeaderIntegrityFetcher interface, with a default implementation that
  uses the above two. Useful in test to stub out the whole thing.

Extracted the computation of CBOR headers and MICE payload from SxgWorker into
utils::signed_headers_and_payload, for reuse in HeaderIntegrityFetcher.

Moved the requirement for status code 200 from index.ts to
signed_headers_and_payload. This also implements the requirement for fastly_compute.

In the fetchers for both cloudflare_worker and fastly_compute, add an 8MB
response body limit, after which an error is returned. In cloudflare_worker,
only attach a request body if it is non-empty; this addresses an error message
that bodies are disallowed on GET requests.

TODO: Issue these subresource fetches concurrently, to speed up processing on
cache miss.

[1] https://github.com/WICG/webpackage/blob/main/explainers/signed-exchange-subresource-substitution.md
[2] https://github.com/google/webpackager/blob/main/docs/cache_requirements.md
[3] https://developer.mozilla.org/en-US/docs/Web/API/Cache
[4] https://developers.cloudflare.com/workers/runtime-apis/cache

/cc @oliy

### CODE REVIEW

File: cloudflare_worker/worker/src/index.ts [hunk 1]
```diff
@@ -263,42 +271,85 @@ async function generateSxgResponse(fallbackUrl: string, certOrigin: string, payl
...
   }
+  let {get: headerIntegrityGet, put: headerIntegrityPut} = await headerIntegrityCache();
   const sxg = await createSignedExchange(
     fallbackUrl,
     certOrigin,
-    payloadStatusCode,
+    payload.status,
     payloadHeaders,
     new Uint8Array(payloadBody),
     Math.round(Date.now() / 1000 - 60 * 60 * 12),
```
Comment (banaag) [inline cloudflare_worker/worker/src/index.ts:288]: Comment on what the magic numbers mean or define a constant.

File: cloudflare_worker/worker/src/index.ts [hunk 2]
```diff
@@ -263,42 +271,85 @@ async function generateSxgResponse(fallbackUrl: string, certOrigin: string, payl
...
       method: request.method,
+  };
+  if (request.body.length > 0) {
+    requestInit.body = new Uint8Array(request.body);
+  }
+  const response = await fetch(request.url, requestInit);
+
+  let body: ArrayBuffer;
+  if (response.body) {
+    const bodyReader = response.body.pipeThrough(limitBytes(8000000));
```
Comment (banaag) [inline cloudflare_worker/worker/src/index.ts:309]: Constant for 8000000

**Review by antiphoton — APPROVED** [review]

**Review by banaag — APPROVED** [review]

File: sxg_rs/src/header_integrity.rs [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    use crate::http_cache::NullCache;
+    use anyhow::{anyhow, Result};
+    use std::collections::HashMap;
+
+    static EMPTY_SET: Lazy<BTreeSet<String>> = Lazy::new(|| BTreeSet::new());
+    static mut NULL_CACHE: NullCache = NullCache {};
+
+    // For use in other modules' tests.
+    pub fn null_integrity_fetcher() -> HeaderIntegrityFetcherImpl<'static, NullFetcher, NullCache> {
+        // Hopefully, the unsafe is OK in test since it's single-threaded.
```
[the remark this answers is not in the document]
Comment (twifkak) [reply sxg_rs/src/header_integrity.rs]: Thanks. Updated the comment.

File: sxg_rs/src/header_integrity.rs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+            &mut cache,
+            &strip_response_headers,
+        );
+        assert_eq!(
+            fetcher.fetch(TEST_URL).await.unwrap(),
+            EXPECTED_HEADER_INTEGRITY
+        );
+    }
+
+    struct InMemoryCache(HashMap<String, HttpResponse>);
```
[the remark this answers is not in the document]
Comment (twifkak) [reply sxg_rs/src/header_integrity.rs:230]: I'd prefer to wait until an actual need, to help dictate requirements. e.g. For production server usage, we'd probably want to add a memory bound w/ an eviction policy. But for some command-line tool, maybe not.

**Comment by twifkak** [comment]
Addresses #13.

**Pull request merged**